File: cmd/gizzi-code/src/scheduler/IntelliScheduleEngine.ts
export interface IntelliScheduleInput {
  tasks: Array<{
    id: string;
    title: string;
    priority: number; // 1-100
    estimatedMinutes: number;
    deadline?: number; // timestamp (ms since epoch)
    dependencies: string[]; // task IDs that must complete first
    tags?: string[];
  }>;
  constraints: {
    availableHoursPerDay: number; // default 8
    startTime: number; // timestamp (ms since epoch)
    bufferMinutes: number; // default 15
  };
}

export interface IntelliScheduleOutput {
  orderedTasks: string[]; // Optimal execution order (task IDs)
  schedule: Array<{
    taskId: string;
    startTime: number; // timestamp
    endTime: number; // timestamp
    risk: 'low' | 'medium' | 'high';
  }>;
  unrunnable: string[]; // Tasks that cannot fit before deadline
}

interface ScheduleEntry {
  taskId: string;
  startTime: number;
  endTime: number;
  risk: 'low' | 'medium' | 'high';
}

export class IntelliScheduleEngine {
  optimize(input: IntelliScheduleInput): IntelliScheduleOutput {
    const { tasks, constraints } = input;
    const dayAllocations: Record<string, number> = {};
    const scheduledMap = new Map<string, ScheduleEntry>();
    const orderedTasks: string[] = [];
    const schedule: ScheduleEntry[] = [];
    const unrunnable: string[] = [];

    // Compute blocking counts (how many tasks depend on each task)
    const blockingCount: Record<string, number> = {};
    for (const task of tasks) {
      blockingCount[task.id] = 0;
    }
    for (const task of tasks) {
      for (const depId of task.dependencies) {
        if (depId in blockingCount) {
          blockingCount[depId]++;
        }
      }
    }

    // Sort by (-blocking_count, deadline_asc, -priority, estimated_duration_asc, id_asc)
    const sorted = [...tasks].sort((a, b) => {
      const bcA = blockingCount[a.id] ?? 0;
      const bcB = blockingCount[b.id] ?? 0;
      if (bcA !== bcB) return bcB - bcA;

      const dlA = a.deadline ?? Infinity;
      const dlB = b.deadline ?? Infinity;
      if (dlA !== dlB) return dlA - dlB;

      if (b.priority !== a.priority) return b.priority - a.priority;

      if (a.estimatedMinutes !== b.estimatedMinutes) {
        return a.estimatedMinutes - b.estimatedMinutes;
      }

      return a.id.localeCompare(b.id);
    });

    const startRef = constraints.startTime;
    const hoursPerDay = constraints.availableHoursPerDay;
    const bufferMs = constraints.bufferMinutes * 60 * 1000;

    if (hoursPerDay <= 0) {
      for (const task of sorted) {
        unrunnable.push(task.id);
      }
      return { orderedTasks, schedule, unrunnable };
    }

    for (const task of sorted) {
      const result = this.allocateTask(
        task,
        startRef,
        hoursPerDay,
        bufferMs,
        dayAllocations,
        scheduledMap,
      );
      if (result) {
        const risk = this.calculateRisk(result.endTime, task.deadline);
        const entry: ScheduleEntry = {
          taskId: task.id,
          startTime: result.startTime,
          endTime: result.endTime,
          risk,
        };
        scheduledMap.set(task.id, entry);
        orderedTasks.push(task.id);
        schedule.push(entry);
      } else {
        unrunnable.push(task.id);
      }
    }

    return { orderedTasks, schedule, unrunnable };
  }

  private allocateTask(
    task: IntelliScheduleInput['tasks'][number],
    startRef: number,
    hoursPerDay: number,
    bufferMs: number,
    dayAllocations: Record<string, number>,
    scheduledMap: Map<string, ScheduleEntry>,
  ): { startTime: number; endTime: number } | null {
    const durationMs = task.estimatedMinutes * 60 * 1000;
    const deadlineMs = task.deadline;

    // Find earliest start considering dependencies
    let earliestStart = startRef;
    for (const depId of task.dependencies) {
      const dep = scheduledMap.get(depId);
      if (dep) {
        earliestStart = Math.max(earliestStart, dep.endTime + bufferMs);
      }
    }

    let remainingMs = durationMs;
    const taskAllocations: Record<string, number> = {};
    let taskStartTime: number | null = null;
    let currentStart = earliestStart;
    let totalDaysChecked = 0;
    const MAX_DAYS = 3650;

    const undoAllocations = () => {
      for (const [day, mins] of Object.entries(taskAllocations)) {
        dayAllocations[day] -= mins;
      }
    };

    while (remainingMs > 0.0001) {
      if (deadlineMs !== undefined && currentStart >= deadlineMs) {
        undoAllocations();
        return null;
      }

      if (totalDaysChecked > MAX_DAYS) {
        undoAllocations();
        return null;
      }
      totalDaysChecked++;

      const dayWorkStart = this.getWorkDayStart(currentStart, startRef);
      const dayWorkEnd = this.getWorkDayEnd(currentStart, startRef, hoursPerDay);
      const dayKey = this.formatDayKey(dayWorkStart);

      if (currentStart >= dayWorkEnd) {
        currentStart = this.getWorkDayStart(
          this.getNextWorkDayStart(currentStart, startRef),
          startRef,
        );
        continue;
      }

      if (currentStart < dayWorkStart) {
        currentStart = dayWorkStart;
      }

      const allocateUntil =
        deadlineMs !== undefined ? Math.min(dayWorkEnd, deadlineMs) : dayWorkEnd;
      const availableInDay = allocateUntil - currentStart;

      if (availableInDay <= 0.0001) {
        currentStart = this.getWorkDayStart(
          this.getNextWorkDayStart(currentStart, startRef),
          startRef,
        );
        continue;
      }

      const wantToAllocate = Math.min(remainingMs, availableInDay);
      const usedMs = (dayAllocations[dayKey] || 0) * 60 * 1000;
      const capacityMs = hoursPerDay * 60 * 60 * 1000;
      const actualAllocateMs = Math.min(
        wantToAllocate,
        Math.max(0, capacityMs - usedMs),
      );

      if (actualAllocateMs > 0.0001) {
        if (taskStartTime === null) {
          taskStartTime = currentStart;
        }
        const addedMinutes = actualAllocateMs / (60 * 1000);
        taskAllocations[dayKey] = (taskAllocations[dayKey] || 0) + addedMinutes;
        dayAllocations[dayKey] = (dayAllocations[dayKey] || 0) + addedMinutes;
        remainingMs -= actualAllocateMs;
        currentStart += actualAllocateMs;
      } else {
        currentStart = this.getWorkDayStart(
          this.getNextWorkDayStart(currentStart, startRef),
          startRef,
        );
      }
    }

    if (taskStartTime !== null) {
      return { startTime: taskStartTime, endTime: currentStart };
    }

    return null;
  }

  private calculateRisk(
    endTime: number,
    deadline?: number,
  ): 'low' | 'medium' | 'high' {
    if (deadline === undefined) return 'low';
    const msUntilDeadline = deadline - endTime;
    const ONE_HOUR = 60 * 60 * 1000;
    if (msUntilDeadline < 24 * ONE_HOUR) return 'high';
    if (msUntilDeadline < 72 * ONE_HOUR) return 'medium';
    return 'low';
  }

  private getWorkDayStart(ms: number, refMs: number): number {
    const ref = new Date(refMs);
    const d = new Date(ms);
    d.setHours(ref.getHours(), ref.getMinutes(), ref.getSeconds(), ref.getMilliseconds());
    return d.getTime();
  }

  private getWorkDayEnd(ms: number, refMs: number, hoursPerDay: number): number {
    return this.getWorkDayStart(ms, refMs) + hoursPerDay * 60 * 60 * 1000;
  }

  private getNextWorkDayStart(ms: number, refMs: number): number {
    const d = new Date(ms);
    d.setDate(d.getDate() + 1);
    return this.getWorkDayStart(d.getTime(), refMs);
  }

  private formatDayKey(ms: number): string {
    const d = new Date(ms);
    const y = d.getFullYear();
    const m = String(d.getMonth() + 1).padStart(2, '0');
    const day = String(d.getDate()).padStart(2, '0');
    return `${y}-${m}-${day}`;
  }
}

File: surfaces/allternit-mobile/ios/Core/IntelliScheduleEngine.swift
import Foundation

// -----------------------------------------------------------------------------
// iOS port of the Intelli-Schedule optimizer.
//
// Source: cmd/gizzi-code/src/scheduler/IntelliScheduleEngine.ts
// Same algorithm: sort tasks by (-blockingCount, deadline, -priority,
// estimatedMinutes, id), then allocate contiguous working time across days
// respecting availableHoursPerDay and dependency buffers.
// -----------------------------------------------------------------------------

/// One schedulable task.
struct IntelliTask: Identifiable, Sendable {
    let id: String
    let title: String
    let priority: Int
    let estimatedMinutes: Int
    let deadline: Date?
    let dependencies: [String]
}

/// A scheduled entry returned by the engine.
struct IntelliScheduleEntry: Sendable {
    let taskId: String
    let startTime: Date
    let endTime: Date
    let risk: IntelliScheduleRisk
}

enum IntelliScheduleRisk: String, Sendable {
    case low
    case medium
    case high
}

/// Constraints passed to the optimizer.
struct IntelliScheduleConstraints: Sendable {
    var availableHoursPerDay: Double = 8
    var startTime: Date = Date()
    var bufferMinutes: Double = 15
}

/// Output of the optimizer.
struct IntelliScheduleOutput: Sendable {
    let orderedTasks: [String]
    let schedule: [IntelliScheduleEntry]
    let unrunnable: [String]
}

/// iOS Intelli-Schedule engine.
final class IntelliScheduleEngine: Sendable {
    func optimize(tasks: [IntelliTask], constraints: IntelliScheduleConstraints) -> IntelliScheduleOutput {
        let hoursPerDay = constraints.availableHoursPerDay
        let bufferMs = constraints.bufferMinutes * 60 * 1_000
        let startRef = constraints.startTime

        var orderedTasks: [String] = []
        var schedule: [IntelliScheduleEntry] = []
        var unrunnable: [String] = []

        guard hoursPerDay > 0 else {
            return IntelliScheduleOutput(orderedTasks: [], schedule: [], unrunnable: tasks.map(\.id))
        }

        // Blocking count: how many tasks depend on each task.
        var blockingCount: [String: Int] = [:]
        for task in tasks {
            blockingCount[task.id] = 0
        }
        for task in tasks {
            for depId in task.dependencies {
                if blockingCount[depId] != nil {
                    blockingCount[depId, default: 0] += 1
                }
            }
        }

        // Sort: more blockers first, then earlier deadline, then higher priority,
        // then shorter estimate, then stable id.
        let sorted = tasks.sorted { a, b in
            let bcA = blockingCount[a.id] ?? 0
            let bcB = blockingCount[b.id] ?? 0
            if bcA != bcB { return bcA > bcB }

            let dlA = a.deadline?.timeIntervalSince1970 ?? Double.infinity
            let dlB = b.deadline?.timeIntervalSince1970 ?? Double.infinity
            if dlA != dlB { return dlA < dlB }

            if a.priority != b.priority { return a.priority > b.priority }
            if a.estimatedMinutes != b.estimatedMinutes {
                return a.estimatedMinutes < b.estimatedMinutes
            }
            return a.id.localizedStandardCompare(b.id) == .orderedAscending
        }

        var scheduledMap: [String: IntelliScheduleEntry] = [:]
        var dayAllocations: [String: Double] = [:]

        for task in sorted {
            guard let result = allocateTask(
                task: task,
                startRef: startRef,
                hoursPerDay: hoursPerDay,
                bufferMs: bufferMs,
                dayAllocations: &dayAllocations,
                scheduledMap: scheduledMap
            ) else {
                unrunnable.append(task.id)
                continue
            }

            let entry = IntelliScheduleEntry(
                taskId: task.id,
                startTime: result.startTime,
                endTime: result.endTime,
                risk: calculateRisk(endTime: result.endTime, deadline: task.deadline)
            )
            scheduledMap[task.id] = entry
            orderedTasks.append(task.id)
            schedule.append(entry)
        }

        return IntelliScheduleOutput(orderedTasks: orderedTasks, schedule: schedule, unrunnable: unrunnable)
    }

    // MARK: - Allocation

    private func allocateTask(
        task: IntelliTask,
        startRef: Date,
        hoursPerDay: Double,
        bufferMs: Double,
        dayAllocations: inout [String: Double],
        scheduledMap: [String: IntelliScheduleEntry]
    ) -> (startTime: Date, endTime: Date)? {
        let durationMs = Double(task.estimatedMinutes) * 60 * 1_000
        let deadlineMs = task.deadline?.timeIntervalSince1970

        // Earliest start respecting dependencies.
        var earliestStart = startRef.timeIntervalSince1970
        for depId in task.dependencies {
            if let dep = scheduledMap[depId] {
                earliestStart = max(earliestStart, dep.endTime.timeIntervalSince1970 + bufferMs / 1_000)
            }
        }

        var remainingMs = durationMs
        var taskAllocations: [String: Double] = [:]
        var taskStartTime: Double?
        var currentStart = earliestStart
        let maxDays = 3650
        var daysChecked = 0

        func dayKey(_ ts: Double) -> String {
            let date = Date(timeIntervalSince1970: ts)
            var cal = Calendar.current
            cal.timeZone = TimeZone.current
            let comps = cal.dateComponents([.year, .month, .day], from: date)
            return "\(comps.year ?? 0)-\(comps.month ?? 0)-\(comps.day ?? 0)"
        }

        func undo() {
            for (day, mins) in taskAllocations {
                dayAllocations[day, default: 0] -= mins
            }
        }

        while remainingMs > 0.0001 {
            if let deadlineMs, currentStart >= deadlineMs {
                undo()
                return nil
            }
            if daysChecked > maxDays {
                undo()
                return nil
            }

            let day = dayKey(currentStart)
            let usedMs = (dayAllocations[day] ?? 0) * 60 * 1_000
            let dayCapacityMs = hoursPerDay * 60 * 60 * 1_000
            let availableMs = max(0, dayCapacityMs - usedMs)

            if taskStartTime == nil {
                taskStartTime = currentStart
            }

            if availableMs <= 0 {
                // Move to next day at start reference hour.
                let currentDate = Date(timeIntervalSince1970: currentStart)
                var cal = Calendar.current
                cal.timeZone = TimeZone.current
                var comps = cal.dateComponents([.year, .month, .day], from: currentDate)
                comps.day = (comps.day ?? 0) + 1
                comps.hour = cal.component(.hour, from: startRef)
                comps.minute = cal.component(.minute, from: startRef)
                comps.second = cal.component(.second, from: startRef)
                if let next = cal.date(from: comps) {
                    currentStart = next.timeIntervalSince1970
                } else {
                    currentStart += 24 * 60 * 60
                }
                daysChecked += 1
                continue
            }

            let takeMs = min(remainingMs, availableMs)
            let takeMinutes = takeMs / (60 * 1_000)
            taskAllocations[day, default: 0] += takeMinutes
            dayAllocations[day, default: 0] += takeMinutes
            remainingMs -= takeMs
            currentStart += takeMs / 1_000
            daysChecked += 1
        }

        guard let start = taskStartTime else { return nil }
        return (
            startTime: Date(timeIntervalSince1970: start),
            endTime: Date(timeIntervalSince1970: currentStart)
        )
    }

    // MARK: - Risk

    private func calculateRisk(endTime: Date, deadline: Date?) -> IntelliScheduleRisk {
        guard let deadline else { return .low }
        let remaining = deadline.timeIntervalSince1970 - endTime.timeIntervalSince1970
        if remaining < 0 { return .high }
        if remaining < 24 * 60 * 60 { return .medium }
        return .low
    }
}
